rpctest/node.go

```diff
@@ -202,16 +206,107 @@ func (n *node) debugf(format string, args ...interface{}) {
 	debugf(n.t, pid+format, args...)
 }
 
+// setTestNodeDir sets the package level testNodeDir. testNodeDirMtx must be
+// held for writes.
+func setTestNodeDir(t *testing.T) error {
+	var err error
+	testNodeDir, err = ioutil.TempDir("", "rpctestdcrdnode")
+	if err != nil {
+		return err
+	}
+	debugf(t, "node dir: %v\n", testNodeDir)
+	return nil
+}
+
+// copyNode will copy the contents of pathToDCRD to testNodeDir/dcrd(.exe).
+// testNodeDirMtx must be held for writes.
+func copyNode(t *testing.T, exe, pathToDCRD string) error {
+	err := setTestNodeDir(t)
+	if err != nil {
+		return err
+	}
+	outputPath := filepath.Join(testNodeDir, exe)
+	if runtime.GOOS == "windows" {
+		outputPath += ".exe"
+	}
+	closeFile := func(f *os.File) {
+		cerr := f.Close()
+		if err == nil {
+			err = cerr
+		}
+	}
+	in, err := os.Open(pathToDCRD)
+	if err != nil {
+		return err
+	}
+	defer closeFile(in)
+	out, err := os.Create(outputPath)
+	if err != nil {
+		return err
+	}
+	defer closeFile(out)
+	if _, err = io.Copy(out, in); err != nil {
+		return err
+	}
+	err = out.Sync()
+	// if nil, err will be populated in a defer if unable to close a file.
+	return err
+}
+
+// buildNode creates a new temporary directory and node and saves the location
+// to a package level variable where it is used for all tests. testNodeDirMtx
+// must be held for writes.
+func buildNode(t *testing.T, exe string) error {
+	err := setTestNodeDir(t)
+	if err != nil {
+		return err
+	}
+	// Determine import path of this package.
+	_, rpctestDir, _, ok := runtime.Caller(1)
+	if !ok {
+		return fmt.Errorf("cannot get path to dcrd source code")
+	}
+	dcrdPkgPath := filepath.Join(rpctestDir, "..", "..")
+	// Build dcrd and output an executable in a static temp path.
+	outputPath := filepath.Join(testNodeDir, exe)
+	if runtime.GOOS == "windows" {
+		outputPath += ".exe"
+	}
+	cmd := exec.Command("go", "build", "-o", outputPath, dcrdPkgPath)
+	err = cmd.Run()
+	if err != nil {
+		return fmt.Errorf("failed to build dcrd: %v", err)
+	}
+	return nil
+}
+
 // newNode creates a new node instance according to the passed config. dataDir
 // will be used to hold a file recording the pid of the launched process, and
-// as the base for the log and data directories for dcrd.
-func newNode(t *testing.T, config *nodeConfig, dataDir string) *node {
+// as the base for the log and data directories for dcrd. If pathToDCRD has a
+// non-zero value, the executable located there is copied to the temporary
+// directory.
+func newNode(t *testing.T, config *nodeConfig, dataDir, pathToDCRD string) (*node, error) {
+	// Create the dcrd node used for tests if not created yet.
+	testNodeDirMtx.Lock()
+	if testNodeDir == "" {
+		var err error
+		if pathToDCRD == "" {
+			err = buildNode(t, config.exe)
+		} else {
+			err = copyNode(t, config.exe, pathToDCRD)
```

Comment:
Why copy the binary instead of setting up config.exe to just use it?

Reply:
I don't know... No longer copying.

Reply:
Better to copy the path to the config? Now always using the global path.